pkg/networking/middleware/retry_middleware.go

```diff
@@ -0,0 +1,108 @@
+package middleware
+
+import (
+	"bytes"
+	"errors"
+	"io"
+	"net/http"
+	"strconv"
+	"time"
+
+	"github.com/cenkalti/backoff/v5"
+
+	"github.com/snyk/go-application-framework/pkg/configuration"
+)
+
+const defaultRetryCount uint = 3
+const configurationKeyRetryCount = "INTERNAL_MAX_NET_RETRIES"
+const defaultRetryAfterSeconds = 5
+
+type RetryMiddleware struct {
+	next   http.RoundTripper
+	config configuration.Configuration
+}
+
+func NewRetryMiddleware(roundTriper http.RoundTripper, config configuration.Configuration) *RetryMiddleware {
+	return &RetryMiddleware{
+		next:   roundTriper,
+		config: config,
+	}
+}
+
+func (rm RetryMiddleware) RoundTrip(req *http.Request) (*http.Response, error) {
+	var finalResponse *http.Response = nil
+	var finalError error = nil
+	var localBodyBuffer []byte
+	var maxRetryCount = defaultRetryCount
+
+	// if a body is available, create a local copy to be able to use it multiple times
+	if req.Body != nil {
+		// possible optimization for large request bodies to not buffer in memory but in filesystem
+		var localBufferError error
+		localBodyBuffer, localBufferError = io.ReadAll(req.Body)
+		if localBufferError != nil {
+			return nil, localBufferError
+		}
+		req.Body = io.NopCloser(bytes.NewBuffer(localBodyBuffer))
+	}
+
+	if rm.config.IsSet(configurationKeyRetryCount) {
+		maxRetryCount = (uint)(rm.config.GetInt(configurationKeyRetryCount))
+	}
+
+	op := func() (*http.Response, error) {
+		// create a local copy of the request
+		localRequest := *req
+		if len(localBodyBuffer) > 0 {
+			localRequest.Body = io.NopCloser(bytes.NewBuffer(localBodyBuffer))
+		}
+
+		// try to send request
+		response, err := rm.next.RoundTrip(&localRequest)
+		if err != nil {
+			return nil, err
+		}
+
+		if retryAfterError := shouldRetry(response); retryAfterError != nil {
```

Comment:
nitpick: `retryAfterError` is confusing here, as you return from errors above. Probably it should just be `retry` as we analyse response status codes.